bin/upgrade.py

```diff
@@ -0,0 +1,248 @@
+                    limits.pop("time_safety_margin")
+
+    if "validator_flags" in data:
+        ...
+        # TODO fix this...
+        # either write a top level testdata.yaml or add it to generators.yaml ?
+
+    timelimit_path = p.path / ".timelimit"
+    if timelimit_path.is_file():
+        if "limits" not in data:
+            data["limits"] = ruamel.yaml.comments.CommentedMap()
+        if "time_limit" in data["limits"]:
+            bar.error(
+                "can't change '.timelimit' file, 'limits.time_limit' already exists in problem.yaml",
+                resume=True,
+            )
+        else:
+            bar.log("change '.timelimit' file to 'limits.time_limit' in problem.yaml")
+            data["limits"]["time_limit"] = float(timelimit_path.read_text())
+            timelimit_path.unlink()
+
+    domjudge_path = p.path / "domjudge-problem.ini"
+    if domjudge_path.is_file():
+        time_limit = None
+        for line in domjudge_path.read_text().splitlines():
+            key, var = map(str.strip, line.strip().split("="))
+            if (var[0] == '"' or var[0] == "'") and (var[-1] == '"' or var[-1] == "'"):
+                var = var[1:-1]
+            if key == "timelimit":
+                time_limit = float(var)
+        if time_limit is not None:
+            if "limits" not in data:
+                data["limits"] = ruamel.yaml.comments.CommentedMap()
+            if "time_limit" in data["limits"]:
+                bar.error(
+                    "can't change '.timelimit' file, 'limits.time_limit' already exists in problem.yaml",
+                    resume=True,
+                )
+            else:
+                bar.log("change 'domjudge-problem.ini' file to 'limits.time_limit' in problem.yaml")
+                data["limits"]["time_limit"] = time_limit
+                domjudge_path.unlink()
+
+    write_yaml(data, p.path / "problem.yaml")
+
+
+def upgrade_testdata_yaml(p: problem.Problem, bar: ProgressBar) -> None:
+    rename = [
+        ("output_validator_flags", "output_validator_args"),
+        ("inut_validator_flags", "inut_validator_args"),
+    ]
+
+    for f in (p.path / "data").rglob("testdata.yaml"):
+        data = read_yaml(f)
+        assert data is not None
+
+        for old, new in rename:
+            if old in data:
+                if new in data:
+                    bar.error(
+                        f"can't change '{old}', '{new}' already exists in {f.relative_to(p.path)}",
+                        resume=True,
+                    )
+                    continue
+                data[new] = data[old]
+                data.pop(old)
+
+        write_yaml(data, f)
+
+
+def _upgrade(p: problem.Problem, bar: ProgressBar) -> None:
+    bar.start(p)
+
+    upgrade_data(p, bar)
+    upgrade_generators_yaml(p, bar)
+    upgrade_statement(p, bar)
+    upgrade_problem_yaml(p, bar)
+    upgrade_testdata_yaml(p, bar)
+    # TODO: output_validators -> output_validator
```

Comment:
:+1:, pending rework of how we handle the asymmetric validators structure (i.e. multiple input/answer validators vs. singular output validator, and maybe the same for visualizers)

Reply:
we already allow only one output validator?

Reply:
Yes, but I mean that currently, the structure is:
```
input_validators/
  validator_1/
    ...files for validator 1...
  validator_2/
    ...files for validator 2...
output_validators/
  only_validator/
    ...files for only validator...
```
But for the output validator, we should change to:
```
output_validator/
  ...files for outout validator...
 ```

Reply:
yes, but I would expect that this does not require a big change in the code?

Reply:
Sure, but we do still need to do this at some point :stuck_out_tongue: I'll resolve this thread for now though, because that rewrite should happen outside this PR :slightly_smiling_face: